### DISCUSSION TITLE
WinUI Community Call: March 6, 2024

### DISCUSSION THREAD

**Discussion by duncanmacmichael** [microsoft/microsoft-ui-xaml#9361] [General]
### Please join us for the March 2024 edition of the WinUI Community Call. This event will be live-streamed to YouTube: https://www.youtube.com/watch?v=uFgGhNGlPm4
#
# Details
Date: Mar 06, 2024
Time: 17:00-18:00 UTC (9:00-10:00am Pacific)

Anyone and everyone is welcome - no pre-registration is required.
This will be an informal interactive live stream directly with members of our engineering team.

## Format
The community call is a call among the WinUI team that is live-streamed onto YouTube. We present on new updates, share information, welcome guests, and answer your questions.

The WinUI and WindowsAppSDK team are back! Sorry we've been quiet; we've heard you're all excited for a call! We've been working on a more sustainable plan to engage with the developer community long term, and we're excited to share more with you all soon over the coming months!

### Agenda
- Intro
- 1.5 releases & updates
- Improved debugging tools in 1.5
- More demos TBD
- Q&A
- More to follow too!
 
### **Q&A Code of Conduct**
- To help ensure Q&A remains beneficial for the majority of the call's developer audience, please consider whether your question could be framed so that it might be more broadly relevant to our larger developer audience by ensuring it is reasonably concise, broadly beneficial, well-suited for the fast-paced live Q&A format, or able to spark open ended discussion.
- Please understand some questions are not well-suited for the fast-paced live Q&A format and should be filed as standalone [Q&A](https://github.com/microsoft/microsoft-ui-xaml/discussions/new?category=q-a&title=%5BYour+Question+Here%5D) discussions where they can be most appropriately handled. Excluding presentation topic requests, this generally includes requests for dedicated follow-ups or technically complex questions that require investigative time or expertise not included in this month's guest list.
 
### Leave us your questions in the comments on this issue or live in the YouTube chat during the stream!

**Comment by d2dyno1** [comment]
What's the expected ETA/release schedule for InkCanvas and InkToolbar?

**Reply by ghost1372** [reply]
they aren't likely to fully land in the 1.5 timeline at this time.
https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md

**Reply by Gaoyifei1011** [reply]
Is segmented control coming in version 1.5?

**Reply by roxk** [reply]
@Gaoyifei1011 Yes. https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#new-selectorbar-control

**Reply by Gaoyifei1011** [reply]
@roxk Thank you very much. I've already seen it in the WinUI 3 Gallery

**Comment by AathifMahir** [comment]
Can we expect General Improvements to Xaml going forward? Xaml been same and never had any innovation for long time now

**Reply by saint4eva** [reply]
Many modern platforms are going for coded UIs - like MVU pattern.

**Reply by llothar** [reply]
Yeah MVU because our software isn't already slow enough.

**Comment by MattBDev** [comment]
How often can we expect community calls in the future? Will we have long gaps like this going forward or was this a rare occurrence while you worked on a plan to engage with us better in the future?

**Comment by christosk92** [comment]
Any plans for supporting WinUI development outside visual studio? Maybe it's technically possible but it's very tied to VS atm with debugging and project templates etc...

**Reply by ReneeGA2020** [reply]
You can do that with vscode. I did.

**Reply by DarranRowe** [reply]
Right now, everything is possible on the command line.
Most of the tooling is fully documented and supported, but the biggest sticking point is the Xaml compiler. You also have to take into account that some of the tools, especially the ones that are part of the Windows SDK, are Windows executables. Anyway, there is an executable version of the Xaml compiler supplied with the Windows App SDK package, but there is a certain implicit feeling that it could be removed at any point. What's more, the only real way to figure out how to use it is by browsing the source code.
Debugging can always be done with WinDbg. Managed code would require the [SOS debugging extension](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/sos-debugging-extension). The project templates are there to set the project up easily, but that doesn't stop you from doing that manually. The issue is that you would need to figure out how to put all of the pieces together.

**Reply by Scottj1s** [reply]
As @DarranRowe mentions, all activities can be performed on the command line.  Windows App SDK has no intrinsic dependencies on Visual Studio, but it does have on MSBuild.  So, for example, a cmake project in VS isn't supported.  But any development environment that supports MSBuild invocation, such as VS Code, should be fine.

**Reply by ghost1372** [reply]
> As @DarranRowe mentions, all activities can be performed on the command line.  Windows App SDK has no intrinsic dependencies on Visual Studio, but it does have on MSBuild.  So, for example, a cmake project in VS isn't supported.  But any development environment that supports MSBuild invocation, such as VS Code, should be fine.

We have to set `<EnableMsixTooling>` in class library if we want to build our app with github action or dotent command

**Reply by christosk92** [reply]
@Scottj1s Thanks for acknowledging this.  
While it is true that WinAppSdk is not intrinsically tied to VisualStudio, we cannot simply use (for example) dotnet build. 
Other UI frameworks like Avalonia allow for the use of a simple, easy to use CLI command. 

While technically possible, it is unnecessarily complicated to figure out all the different flags of the different compilers (msbuild, xaml) we have to use to compile a WinUI application thus forcing most developers to use VisualStudio.

**Reply by christosk92** [reply]
To give a more concrete comment, I would really love to see official support for Jetbrains Rider.

**Reply by DarranRowe** [reply]
@christosk92 
As @ghost1372 rightfully stated, you have to add EnableMsixTooling to the project, but once everything has been set up, dotnet build works.

**Comment by ReneeGA2020** [comment]
1. How about making C# WinUI template a standard component in the vs desktop workload instead of optional?
2. Making DI and MVVM dependency by default to the template?
3. Making the template to support .net 8/9, Making it by default ImplicitUsings and Nullable?
4. Property window to display the property for the xaml controls? MAUI has that.
5. xaml Grid RowDefinitions="10,10" auto completion support? MAUI has that.
6. Better Windowing support, like WinUIEx did? It's hard to operate windows, while the project is called WINDOWS UI. I still struggle to figure out how to make a window modal.
7. Any progress for the .NET NativeAOT support? Last time I check, it needs CsWinRT to support NativeAOT, any update since then?

**Reply by DarranRowe** [reply]
For 1, isn't that a VS thing?

As far as 3 is concerned, the biggest issue was that the .NET RID change happened right in the middle of the 1.4 development cycle, and .NET 8 was due to be released after 1.4. The team chose to not support .NET 8 in 1.4 because of this.
It has been stated on the Windows App SDK repo that 1.5 will support the changes for .NET 8. This is the biggest block right now. What's more, if you look at the actual project templates, the 1.5p1 templates have the changes in place already. The [publish profiles](https://github.com/microsoft/WindowsAppSDK/tree/v1.5.0-preview1/dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/Properties/PublishProfiles) have been renamed. The [.pubxml](https://github.com/microsoft/WindowsAppSDK/blob/v1.5.0-preview1/dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/Properties/PublishProfiles/win-x64.pubxml) files themselves check for .NET 8 or newer and set the RID based on that. The same is true with the [.csproj file](https://github.com/microsoft/WindowsAppSDK/blob/v1.5.0-preview1/dev/VSIX/ProjectTemplates/Desktop/CSharp/SingleProjectPackagedApp/ProjectTemplate.csproj). Finally, the Windows App SDK NuGet package has RID directories that has been doubled up.

![Screenshot 2024-02-27 144250](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/9310af82-b044-4a64-afae-57be1f417026)

That should be most of the build fixes, and I'm sure that if there are any remaining issues, they will be in place by 1.5.

**Reply by Scottj1s** [reply]
Regarding 1 and 3 (templates):
The Visual Studio team is very careful about what extensions are included in the installer (vs downloaded from the Marketplace), and even more so, what features are on by default.  
Yes, the work has been done to support .NET 8.  Our hope is to get this integrated into the next release of Visual Studio, 17.10.

Regarding 7 (NativeAOT):
We're making progress and plan to make this available in Windows App SDK 1.6

**Reply by HEIC-to-JPEG-Dev** [reply]
I think on the website there should have been a large red & yellow banner "WARNING - .Net 8 is not supported". I upgraded to .Net as soon as it was released, but a message on WASDK saying it wasn't supported must have been buried somewhere.

**Reply by DarranRowe** [reply]
It is supported in 1.5.
The issue here is that the project templates are currently lagging behind because of how they are distributed. You can use the templates currently [available online](https://github.com/microsoft/WindowsAppSDK/tree/v1.5.0-preview1/dev/VSIX/ProjectTemplates/Desktop/CSharp) as a reference.
The last release of 1.4 did also have some of the fixes, it would warn about the RIDs while building.

**Comment by omikhailov** [comment]
What can you say about the following:

![LinkedIn](https://github.com/microsoft/microsoft-ui-xaml/assets/75426711/53f8f40f-0266-4ed3-a67b-8d0014a71fc1)

**Reply by MEK3DK** [reply]
Why would there be any jobs related to WinUI? It is not ready for real production use, thus companies does not hire people for it. 
I think it is 3 years ago we did the first investigation into WinUI 3 to see if it could be used, but quickly concluded that it was no way near ready. Some things have improved, but realistically it is only for hobby use (although the missing designer limits hobby developer too).

**Reply by wbokkers** [reply]
@MEK3DK Our Xelion Windows App is a WinUI 3 app and is widely used. Currently our app has a very low crash rate.  Last time I checked I received quite some money for adding features to this app, so no it's not a hobby.  ;)
See: https://softphone.xelion.com

**Reply by NicholasChrzan** [reply]
I don't believe that WinUI3 isn't ready for production use. It's just deficient compared with the existing functionality of WPF. The reality is that WinUI3 is mostly an "upgrade" path for UWP based applications (if you can work around the various missing pieces of functionality). We moved to it because it provided multi-window support, removed the sandboxing, and gave full access to the Win32 and other C based APIs. This allowed us to combine our UWP app with a WPF application into a single application. The transition from WPF to WinUI3 was very painful (things like min/max window size and owner drawn windows were very painful) and required a great deal of interfacing with the underlying win32 APIs to recreate things that we base properties in WPF. If our UWP based application wasn't an order of magnitude larger and more complex than our WPF app we would have migrated it to WPF instead of WinUI3. 

Until they reduce the number of deficiencies between WPF window support and the current state of WinUI3 windows it's going to be very hard to convince any team currently using WPF to port the app or start new development in WinUI3. It's simply too difficult to replicate functionality that were simple properties of the Window in WPF. In my opinion they should be focusing on improving this and working on bug fixes rather than new functionality and feature. What is currently available in WinUI3 is already pretty rich feature wise, but it simply isn't bulletproof in all scenarios and requires custom workarounds to get around issues.

**Reply by wbokkers** [reply]
@NicholasChrzan Yes, you are right that a fair amount of low level win32 APIs are needed. For me that is not a deal-breaker, but I can imagine this is painful coming from a WPF background. Libraries (like [WinUIEx](https://github.com/dotMorten/WinUIEx)) can be of great help. I already had my own windowing code in place, before WinUIEx existed, but I don't mind fiddling with win32.

**Reply by pjmlp** [reply]
If only WinUI3  was actually an upgrade, instead of a downgrade without tooling parity to the designer, .NET Native, C++/CX, and features years away, assuming Microsoft doesn't eventually replace everything with Blazor.

**Reply by LudoTexx** [reply]
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/19667047/8f23fe2c-821b-4cf5-bf63-d594e917028b)

Well, you can found one at least, everything needs a start someday... (Coca-Cola)

**Comment by ghord** [comment]
1. Any plans to address Drag&Drop issues in elevated apps? At least an option to globally disable Drag&Drop would be helpful.
2. Is proper support of `ShowDialog` for `Window` object planned? Currently emulating modal windows using `SetWindowLongPtr(GWLP_HWNDPARENT, ...)` and disabling parent window breaks `Popup` and `ComboBox` controls in child modal window. This is must have api for multi-window applications, really weird that it is missing.
3. Anything on keyboard support for Date Picker?

**Reply by rlwakefield** [reply]
Yes please! This functionality gap is something that is holding back the Terminal team from adding support for drag n drop re-ordering of the tabs in an elevated session. It works in the normal instance, but not elevated. Please fix this!

**Reply by DarranRowe** [reply]
>Currently emulating modal windows using SetWindowLongPtr(GWLP_HWNDPARENT, ...) and disabling parent window breaks Popup and ComboBox controls in child modal window.

Not to mention that SetWindowLongPtr using GWLP_HWNDPARENT is undocumented and completely unsupported in Windows itself.

**Reply by wbokkers** [reply]
Yes, modal window support would be much appreciated. Using SetWindowLongPtr feels as a terrible hack to me.

**Comment by Rarisma** [comment]
Any updates on going open source?

**Reply by jamespack** [reply]
Isn't it already?

**Reply by Scottj1s** [reply]
We've been careful to use the term "source available", which, with public debug symbols, enables debugging the Windows App SDK binaries.  Full "open source", with the expectation of buildability, testability, accepting PRs, and so on, is still work in progress.  No updates on that.

**Reply by jamespack** [reply]
> We've been careful to use the term "source available", which, with public debug symbols, enables debugging the Windows App SDK binaries.  Full "open source", with the expectation of buildability, testability, accepting PRs, and so on, is still work in progress.  No updates on that. 

That's good enough for most use cases I bet.

**Comment by ghost1372** [comment]
Any news about Creating a Windows 11 context menu without using C++ and IExplorerCommand?
There is an open issue (Feature Request) for this from 3 years ago.
 https://github.com/microsoft/WindowsAppSDK/issues/1082

**Comment by jschwizer99** [comment]
There was good progress with the last few releases in respect of bug fixes. The `TreeView` control was introduced very early in the WinUI development and didn't get a lot of improvements. Especially, the multi-select functionality suffers from a lot of issues. In our productive code, this control type is responsible for countless issues. We had to disable the multi-select functionality as it strongly affected application stability. When compared to the `ListView` or other controls for collections the `TreeView` is still in its infancy.

Are there any plans to improve the `TreeView` control and fix the bugs related to selection and virtualization?

Here is a long list of issues still open on `TreeView` #8887, #8791, #8726, #8298, #8280, #8282, #8275, #8171 ... and many more

**Reply by codendone** [reply]
Yes, there are plans to look at top issues of TreeView during 1.6. I'm not sure the set of issues has been selected yet, but I added your comment to the internal task.

**Comment by Gaoyifei1011** [comment]
> ### Please join us for the March 2024 edition of the WinUI Community Call. This event will be live-streamed to YouTube: https://www.youtube.com/watch?v=uFgGhNGlPm4
> # Details
> Date: Mar 06, 2024 Time: 17:00-18:00 UTC (9:00-10:00am Pacific)
> 
> Anyone and everyone is welcome - no pre-registration is required. This will be an informal interactive live stream directly with members of our engineering team.
> 
> ## Format
> The community call is a call among the WinUI team that is live-streamed onto YouTube. We present on new updates, share information, welcome guests, and answer your questions.
> 
> The WinUI and WindowsAppSDK team are back! Sorry we've been quiet; we've heard you're all excited for a call! We've been working on a more sustainable plan to engage with the developer community long term, and we're excited to share more with you all soon over the coming months!
> 
> ### Agenda
> * Intro
> * 1.5 releases & updates
> * Improved debugging tools in 1.5
> * More demos TBD
> * Q&A
> * More to follow too!
> 
> ### **Q&A Code of Conduct**
> * To help ensure Q&A remains beneficial for the majority of the call's developer audience, please consider whether your question could be framed so that it might be more broadly relevant to our larger developer audience by ensuring it is reasonably concise, broadly beneficial, well-suited for the fast-paced live Q&A format, or able to spark open ended discussion.
> * Please understand some questions are not well-suited for the fast-paced live Q&A format and should be filed as standalone [Q&A](https://github.com/microsoft/microsoft-ui-xaml/discussions/new?category=q-a&title=%5BYour+Question+Here%5D) discussions where they can be most appropriately handled. Excluding presentation topic requests, this generally includes requests for dedicated follow-ups or technically complex questions that require investigative time or expertise not included in this month's guest list.
> 
> ### Leave us your questions in the comments on this issue or live in the YouTube chat during the stream!

I'd like to know when WinUI team will WinUI 3 UIElement. ProtectedCursor properties change from protected into public?
https://learn.microsoft.com/zh-cn/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.uielement.protectedcursor?view=windows-app-sdk-1.4

**Comment by lhak** [comment]
There a quite a few APIs that have been in the experimental releases for a long time but have not been promoted to the stable release. These includes the dpi related functionality in the DisplayInformation class (currently win32 interop code is required here) and ContentExternalOutputLink class (for blending additional content with swap chains which is otherwise not possible at all). When can we expect these APIs to be available in a stable release? Alternatively, maybe some functionality could be introduced that let you enable experimental APIs in stable releases.

**Reply by ghord** [reply]
Not tried it, but shouldn't you be able to look at sources of Experimental release and try to do the same thing as they do with `ContentExternalOutputLink` in your project?

**Reply by lhak** [reply]
The source code for these APIs is not public

**Comment by HEIC-to-JPEG-Dev** [comment]
A Visual XAML editor inside VS is the in the Top 5 voted for things, any thoughts on this ? bringing it on-par with WinForms, WPF, and UWP.

**Comment by ghost1372** [comment]
WPF Team is introducing WinUI 3 styles and controls for WPF, isn't this a threat to WinUI3?

**Reply by Gaoyifei1011** [reply]
I think not, Windowsappsdk will soon support native aot, in the binary distribution volume, Windowsappsdk will have more advantages

**Reply by ghost1372** [reply]
WPF can run on Windows 7+, I think many developers prefer to use it.

**Reply by HEIC-to-JPEG-Dev** [reply]
WindowsAppSDK works on WPF (it's a promoted feature) - so won't that capability come to WPF as well ?

I have to admit, with WPF using .net 8 & WASDK 1.5, it compiles faster, has an in-VS visual editor, property lists, and looks the same (with WPF-UI, which is an add-on, but then you need things like WinUIEx for WinUI). Don't get me wrong, my apps are 100% WinUI 3 Desktop, but I wouldn't disregard WPF.

**Reply by ReneeGA2020** [reply]
Currently I still prefer WinUI as it has the features of x:bind and SwapChainPanel. That being said, if WPF got those features, I will choice WPF, as WPF has a designer.

**Comment by lolametro** [comment]
Was the blog post/documentation regarding "Improved debugging tools" that @Scottj1s presented on the calll already published? I was not able to find it, but I think it would highly benefit devs.

**Reply by Scottj1s** [reply]
@lolametro thanks for the suggestion - we'll consider adding that content to the [Windows App SDK docs](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/).

**Comment by JeganRavichandran** [comment]
In WinUI, can we use a C# library/component in C++ Application?

https://github.com/microsoft/microsoft-ui-xaml/issues/7294
https://github.com/microsoft/microsoft-ui-xaml/issues/9606
https://github.com/microsoft/microsoft-ui-xaml/issues/9458
